client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +291,54 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+    def processResponse(
+        response: Response[F],
+        disposed: Ref[F, Boolean],
+    ): Resource[F, Response[F]] =
+      for {
+        channel <- Resource.eval(Channel.synchronous[F, Chunk[Byte]])
+
+        _ <- response.body.chunks
+          .through(channel.sendAll)
+          .compile
+          .drain
+          .uncancelable
+          .background
+
+        r = response.withBodyStream(
+          Stream
+            .eval(disposed.get)
+            .ifM(
+              Stream.raiseError[F](new IOException("response was disposed")),
+              channel.stream.unchunks
+                .onFinalize(
+                  channel.stream.compile.drain
+                    .guarantee(disposed.set(true))
```

Comment:
Is this setting `disposed` to `true` a thing if we run the stream after the final check is performed (322-326 lines of this PR)?

Reply:
Yep. It happens becuase the final check uses the body's stream that contains the `disposed` setting rather than the `channel.stream`.